File: python/src/warehouse/warehouse_desk_app.py
class WarehouseDeskApp:
    def __init__(self):
        self._stock: dict[str, int] = {}
        self._reserved: dict[str, int] = {}
        self._price: dict[str, float] = {}
        self._order_status: dict[str, str] = {}
        self._order_sku: dict[str, str] = {}
        self._order_qty: dict[str, int] = {}
        self._event_log: list[str] = []
        self._cash_balance: float = 0.0
        self._next_order_number: int = 1001

    def seed_data(self):
        self._stock = {"PEN-BLACK": 40, "PEN-BLUE": 25, "NOTE-A5": 15, "STAPLER": 4}
        self._reserved = {"PEN-BLACK": 0, "PEN-BLUE": 0, "NOTE-A5": 0, "STAPLER": 0}
        self._price = {"PEN-BLACK": 1.5, "PEN-BLUE": 1.6, "NOTE-A5": 4.0, "STAPLER": 12.0}
        self._cash_balance = 300.0
        self._next_order_number = 1001

    def run_demo_day(self):
        commands = [
            "RECV;NOTE-A5;5;2.20",
            "SELL;alice;PEN-BLACK;10",
            "SELL;bob;STAPLER;5",
            "CANCEL;O1002",
            "COUNT;STAPLER",
            "SELL;carol;STAPLER;2",
            "SELL;dan;NOTE-A5;14",
            "COUNT;NOTE-A5",
            "DUMP",
        ]
        for command in commands:
            self.process_line(command)
        self.print_end_of_day_report()

    def process_line(self, line: str):
        parts = line.split(";")
        cmd = parts[0]

        if cmd == "RECV":
            sku, qty, unit_cost = parts[1], int(parts[2].strip()), float(parts[3].strip())
            self._stock[sku] = self._stock.get(sku, 0) + qty
            self._cash_balance -= qty * unit_cost
            self._event_log.append(f"received {qty} of {sku} at {unit_cost}")
            return

        if cmd == "SELL":
            customer, sku, qty = parts[1], parts[2], int(parts[3].strip())
            order_id = f"O{self._next_order_number}"
            self._next_order_number += 1
            self._order_sku[order_id] = sku
            self._order_qty[order_id] = qty

            on_hand = self._stock.get(sku, 0)
            reserved = self._reserved.get(sku, 0)
            available = on_hand - reserved
            if available < qty:
                self._order_status[order_id] = "BACKORDER"
                self._event_log.append(f"order {order_id} backordered for {customer} sku={sku} qty={qty}")
            else:
                self._stock[sku] = on_hand - qty
                order_total = self._price.get(sku, 0.0) * qty
                self._cash_balance += order_total
                self._order_status[order_id] = "SHIPPED"
                self._event_log.append(f"order {order_id} shipped to {customer} amount={order_total}")
            return

        if cmd == "CANCEL":
            order_id = parts[1]
            status = self._order_status.get(order_id)
            if status is None:
                self._event_log.append(f"cannot cancel {order_id} because it does not exist")
                return
            if status == "BACKORDER":
                self._order_status[order_id] = "CANCELLED"
                self._event_log.append(f"cancelled backorder {order_id}")
                return
            if status == "SHIPPED":
                sku = self._order_sku[order_id]
                qty = self._order_qty.get(order_id, 0)
                self._stock[sku] = self._stock.get(sku, 0) + qty
                self._cash_balance -= self._price.get(sku, 0.0) * qty
                self._order_status[order_id] = "CANCELLED_AFTER_SHIP"
                self._event_log.append(f"cancelled shipped order {order_id} with restock")
                return
            self._event_log.append(f"order {order_id} could not be cancelled from state {status}")
            return

        if cmd == "COUNT":
            sku = parts[1]
            on_hand = self._stock.get(sku, 0)
            reserved = self._reserved.get(sku, 0)
            available = on_hand - reserved
            self._event_log.append(f"count {sku} onHand={on_hand} reserved={reserved} available={available}")
            return

        if cmd == "DUMP":
            print("---- dump ----")
            print(f"stock={self._stock}")
            print(f"reserved={self._reserved}")
            print(f"orders={self._order_status}")
            print(f"cashBalance={self._cash_balance}")
            return

        self._event_log.append(f"unknown command: {line}")

    def print_end_of_day_report(self):
        shipped = sum(1 for s in self._order_status.values() if s == "SHIPPED")
        backorder = sum(1 for s in self._order_status.values() if s == "BACKORDER")
        cancelled = sum(1 for s in self._order_status.values() if s.startswith("CANCELLED"))
        low_stock = [sku for sku, qty in self._stock.items() if qty < 5]

        print()
        print("==== end of day ====")
        print(f"orders shipped: {shipped}")
        print(f"orders backordered: {backorder}")
        print(f"orders cancelled: {cancelled}")
        print(f"cash balance: {self._cash_balance:.2f}")
        print(f"low stock skus: {low_stock}")
        print()
        print("events:")
        for event in self._event_log:
            print(f" - {event}")

File: csharp/Warehouse-Desk-App/Warehouse-Desktop/WarehouseDeskApp.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Warehouse_Desktop
{
    public class WarehouseDeskApp
    {
        private readonly Dictionary<string, int> stockBySku = new Dictionary<string, int>();
        private readonly Dictionary<string, int> reservedBySku = new Dictionary<string, int>();
        private readonly Dictionary<string, double> priceBySku = new Dictionary<string, double>();
        private readonly Dictionary<string, string> orderStatus = new Dictionary<string, string>();
        private readonly Dictionary<string, string> orderSku = new Dictionary<string, string>();
        private readonly Dictionary<string, int> orderQty = new Dictionary<string, int>();
        private readonly Dictionary<string, string> reservationSku = new Dictionary<string, string>();
        private readonly Dictionary<string, int> reservationQty = new Dictionary<string, int>();
        private readonly Dictionary<string, DateTime> reservationExpiry = new Dictionary<string, DateTime>();
        private readonly Dictionary<string, string> reservationCustomer = new Dictionary<string, string>();
        private readonly List<string> eventLog = new List<string>();
        public IReadOnlyList<string> EventLog => eventLog.AsReadOnly();
        private double cashBalance;
        private int nextOrderNumber;

        public void SeedData()
        {
            stockBySku["PEN-BLACK"] = 40;
            stockBySku["PEN-BLUE"] = 25;
            stockBySku["NOTE-A5"] = 15;
            stockBySku["STAPLER"] = 4;

            reservedBySku["PEN-BLACK"] = 0;
            reservedBySku["PEN-BLUE"] = 0;
            reservedBySku["NOTE-A5"] = 0;
            reservedBySku["STAPLER"] = 0;

            priceBySku["PEN-BLACK"] = 1.5;
            priceBySku["PEN-BLUE"] = 1.6;
            priceBySku["NOTE-A5"] = 4.0;
            priceBySku["STAPLER"] = 12.0;

            cashBalance = 300.0;
            nextOrderNumber = 1001;
        }

        public void RunDemoDay()
        {
            var commands = new List<string>
            {
                "RECV;NOTE-A5;5;2.20",
                "SELL;alice;PEN-BLACK;10",
                "SELL;bob;STAPLER;5",
                "CANCEL;O1002",
                "COUNT;STAPLER",
                "SELL;carol;STAPLER;2",
                "SELL;dan;NOTE-A5;14",
                "COUNT;NOTE-A5",
                "DUMP"
            };

            foreach (var command in commands)
            {
                ProcessLine(command);
            }
            PrintEndOfDayReport();
        }

        public void ProcessLine(string line)
        {
            ProcessLine(line, DateTime.Now);
        }

        public void ProcessLine(string line, DateTime currentTime)
        {
            ExpireReservations(currentTime);
            string[] parts = line.Split(';');
            string type = parts[0];

            if ("RECV".Equals(type))
            {
                string sku = parts[1];
                int qty = ParseInt(parts[2]);
                double unitCost = ParseDouble(parts[3]);
                int current = stockBySku.GetValueOrDefault(sku, 0);
                stockBySku[sku] = current + qty;
                cashBalance = cashBalance - (qty * unitCost);
                eventLog.Add("received " + qty + " of " + sku + " at " + unitCost);
                return;
            }

            if ("SELL".Equals(type))
            {
                string customer = parts[1];
                string sku = parts[2];
                int qty = ParseInt(parts[3]);
                string orderId = "O" + nextOrderNumber;
                nextOrderNumber = nextOrderNumber + 1;
                orderSku[orderId] = sku;
                orderQty[orderId] = qty;

                int onHand = stockBySku.GetValueOrDefault(sku, 0);
                int reserved = reservedBySku.GetValueOrDefault(sku, 0);
                int available = onHand - reserved;
                if (available < qty)
                {
                    orderStatus[orderId] = "BACKORDER";
                    eventLog.Add("order " + orderId + " backordered for " + customer + " sku=" + sku + " qty=" + qty);
                }
                else
                {
                    stockBySku[sku] = onHand - qty;
                    double unitPrice = priceBySku.GetValueOrDefault(sku, 0.0);
                    double orderTotal = unitPrice * qty;
                    cashBalance = cashBalance + orderTotal;
                    orderStatus[orderId] = "SHIPPED";
                    eventLog.Add("order " + orderId + " shipped to " + customer + " amount=" + orderTotal);
                }
                return;
            }

            if ("CANCEL".Equals(type))
            {
                string orderId = parts[1];
                if (!orderStatus.TryGetValue(orderId, out var status))
                {
                    eventLog.Add("cannot cancel " + orderId + " because it does not exist");
                    return;
                }

                if ("BACKORDER".Equals(status))
                {
                    orderStatus[orderId] = "CANCELLED";
                    eventLog.Add("cancelled backorder " + orderId);
                    return;
                }

                if ("SHIPPED".Equals(status))
                {
                    string sku = orderSku[orderId];
                    int qty = orderQty.GetValueOrDefault(orderId, 0);
                    int current = stockBySku.GetValueOrDefault(sku, 0);
                    stockBySku[sku] = current + qty;
                    double unitPrice = priceBySku.GetValueOrDefault(sku, 0.0);
                    cashBalance = cashBalance - (unitPrice * qty);
                    orderStatus[orderId] = "CANCELLED_AFTER_SHIP";
                    eventLog.Add("cancelled shipped order " + orderId + " with restock");
                    return;
                }

                eventLog.Add("order " + orderId + " could not be cancelled from state " + status);
                return;
            }

            if ("COUNT".Equals(type))
            {
                string sku = parts[1];
                int onHand = stockBySku.GetValueOrDefault(sku, 0);
                int reserved = reservedBySku.GetValueOrDefault(sku, 0);
                int available = onHand - reserved;
                eventLog.Add("count " + sku + " onHand=" + onHand + " reserved=" + reserved + " available=" + available);
                return;
            }

            if ("RESERVE".Equals(type))
            {
                string customer = parts[1];
                string sku = parts[2];
                int qty = ParseInt(parts[3]);
                int minutes = ParseInt(parts[4]);

                int onHand = stockBySku.GetValueOrDefault(sku, 0);
                int reserved = reservedBySku.GetValueOrDefault(sku, 0);
                int available = onHand - reserved;

                if (available < qty)
                {
                    eventLog.Add("cannot reserve " + qty + " of " + sku + " for " + customer + ": insufficient stock");
                }
                else
                {
                    string resId = "R" + nextOrderNumber;
                    nextOrderNumber++;
                    reservationSku[resId] = sku;
                    reservationQty[resId] = qty;
                    reservationCustomer[resId] = customer;
                    reservationExpiry[resId] = currentTime.AddMinutes(minutes);

                    reservedBySku[sku] = reserved + qty;
                    eventLog.Add("reserved " + qty + " of " + sku + " for " + customer + " (id=" + resId + ")");
                }
                return;
            }

            if ("CONFIRM".Equals(type))
            {
                string resId = parts[1];
                if (reservationSku.TryGetValue(resId, out string? sku))
                {
                    int qty = reservationQty[resId];
                    string customer = reservationCustomer[resId];

                    // Ship it
                    stockBySku[sku] -= qty;
                    reservedBySku[sku] -= qty;

                    string orderId = "O" + nextOrderNumber++;
                    orderSku[orderId] = sku;
                    orderQty[orderId] = qty;
                    orderStatus[orderId] = "SHIPPED";

                    double unitPrice = priceBySku.GetValueOrDefault(sku, 0.0);
                    double orderTotal = unitPrice * qty;
                    cashBalance += orderTotal;

                    // Remove reservation
                    RemoveReservation(resId);

                    eventLog.Add("reservation " + resId + " confirmed and shipped as " + orderId);
                }
                else
                {
                    eventLog.Add("cannot confirm " + resId + ": reservation expired or not found");
                }
                return;
            }

            if ("RELEASE".Equals(type))
            {
                string resId = parts[1];
                if (reservationSku.TryGetValue(resId, out string? sku))
                {
                    int qty = reservationQty[resId];
                    reservedBySku[sku] -= qty;
                    RemoveReservation(resId);
                    eventLog.Add("reservation " + resId + " released");
                }
                else
                {
                    eventLog.Add("cannot release " + resId + ": reservation expired or not found");
                }
                return;
            }

            if ("DUMP".Equals(type))
            {
                Console.WriteLine("---- dump ----");
                Console.WriteLine("stock={" + string.Join(", ", stockBySku.Select(kvp => kvp.Key + "=" + kvp.Value)) + "}");
                Console.WriteLine("reserved={" + string.Join(", ", reservedBySku.Select(kvp => kvp.Key + "=" + kvp.Value)) + "}");
                Console.WriteLine("orders={" + string.Join(", ", orderStatus.Select(kvp => kvp.Key + "=" + kvp.Value)) + "}");
                Console.WriteLine("cashBalance=" + cashBalance);
                return;
            }

            eventLog.Add("unknown command: " + line);
        }

        private void ExpireReservations(DateTime currentTime)
        {
            var expiredIds = reservationExpiry
                .Where(kvp => kvp.Value <= currentTime)
                .Select(kvp => kvp.Key)
                .ToList();

            foreach (var resId in expiredIds)
            {
                string sku = reservationSku[resId];
                int qty = reservationQty[resId];
                reservedBySku[sku] -= qty;
                RemoveReservation(resId);
                eventLog.Add("reservation " + resId + " expired");
            }
        }

        private void RemoveReservation(string resId)
        {
            reservationSku.Remove(resId);
            reservationQty.Remove(resId);
            reservationCustomer.Remove(resId);
            reservationExpiry.Remove(resId);
        }

        private int ParseInt(string value)
        {
            return int.Parse(value.Trim());
        }

        private double ParseDouble(string value)
        {
            return double.Parse(value.Trim());
        }

        public void PrintEndOfDayReport()
        {
            int shipped = 0;
            int backorder = 0;
            int cancelled = 0;
            foreach (string status in orderStatus.Values)
            {
                if ("SHIPPED".Equals(status))
                {
                    shipped = shipped + 1;
                }
                else if ("BACKORDER".Equals(status))
                {
                    backorder = backorder + 1;
                }
                else if (status.StartsWith("CANCELLED"))
                {
                    cancelled = cancelled + 1;
                }
            }

            List<string> lowStock = new List<string>();
            foreach (var item in stockBySku)
            {
                if (item.Value < 5)
                {
                    lowStock.Add(item.Key);
                }
            }

            Console.WriteLine();
            Console.WriteLine("==== end of day ====");
            Console.WriteLine("orders shipped: " + shipped);
            Console.WriteLine("orders backordered: " + backorder);
            Console.WriteLine("orders cancelled: " + cancelled);
            Console.WriteLine("cash balance: " + string.Format("{0:F2}", cashBalance));
            Console.WriteLine("low stock skus: [" + string.Join(", ", lowStock) + "]");
            Console.WriteLine();
            Console.WriteLine("events:");
            foreach (string @event in eventLog)
            {
                Console.WriteLine(" - " + @event);
            }
        }
    }
}
